content/2019.04.getting-started-with-elm.rst

```diff
@@ -0,0 +1,784 @@
+provided by the platform:
+
+.. code-block:: bash
+
+    ~/tutorial$ elm reactor
+    Go to <http://localhost:8000> to see your project dashboard.
+
+Then open http://localhost:8000/src/Main.elm in your favorite Web browser.
+
+
+Learning about the Elm virtual DOM
+==================================
+
+Virtual DOM functions to generate HTML are in the ``Html`` module.
+
+The ``Html`` module we used above to render some text also exposes many
+more functions for rendering HTML tags. You can import them all using:
+
+.. code-block:: elm
+
+    import Html exposing (..)
+
+Note that unlike with some other languages, the Elm compiler will
+complain if you try to import symbols already defined in the current
+module, which makes it actually useful and really enjoyable to use.
+
+You can then use ``text`` directly for instance:
+
+.. code-block:: elm
+
+    main = text "Hello world"
+
+The Virtual DOM HTML nodes are functions named after standard HTML
+tags, and take two parameters:
+
+- A list of attributes
+- A list of children
+
+If I want to create a ``div`` with a link it would look like this:
+
+.. code-block:: elm
+
+    module Main exposing (main)
+
+    import Html exposing (..)
+    import Html.Attributes exposing (..)
+
+
+    main =
+        div
+            [ class "container" ]
+            [ a
+                [ href "http://www.servicedenuages.fr/" ]
+                [ text "Blog" ]
+            ]
+
+We can also create a list of links in our div:
+
+.. code-block:: elm
+
+    module Main exposing (main)
+
+    import Html exposing (..)
+    import Html.Attributes exposing (..)
+
+
+    main =
+        div
+            [ class "container" ]
+            [ ul
+                [ class "links" ]
+                [ li
+                    []
+                    [ a
+                        [ href "http://www.servicedenuages.fr/" ]
+                        [ text "Blog" ]
+                    ]
+                , li
+                    []
+                    [ a
+                        [ href "http://www.elm-lang.org/" ]
+                        [ text "Elm lang" ]
+                    ]
+                ]
+            ]
+
+
+Adding some state
+=================
+
+Now that you know how to render your page in HTML, let's see how to
+write a program that handles events.
+
+The way Elm handles that is by having:
+
+- a ``Model``, an Elm record: a bit like a JavaScript object with
+  properties, that keep the state of the app
+- an update function that will handle all the app events and update
+  the model state accordingly
+- a view function that will return the Virtual DOM that matches the state
+  every time it's updated.
+
+For those who know Redux_, it has been heavily inspired by
+Elm. Basically ``update`` is a reducer.
+
+.. _Redux: http://redux.js.org/
+
+The events and their parameters are defined in a ``Msg`` type, which
+is a bit like a enum that would take parameters.
```

Comment:
Again this seems like it's combining discussion of language features with architectural structure.